### ISSUE TITLE
VIP: Standardize compiler output artifacts

### ISSUE THREAD

**Issue by fubuloubu** [ethereum/vyper#642]
## Preamble

VIP: <to be assigned>
Title: Standardize compiler output artifacts
Author: @fubuloubu
Type: Standard
Status: Draft
Created: 2018-01-22

## Simple Summary
Change default compiler output to match solc for compatibility with existing/future tools

## Abstract
The solidity compiler outputs a JSON-formatted string containing the compilation artifacts for all compiled contracts. The format is:
```python
{
  "contracts" : {
    "path/to/contractA.ext" : {
      "abi" : "[\" ... \"]", # Not sure why this is in quotes actually...
      "bin" : "0x...",
      "bin-runtime" : "0x...",
      ... # Additional stuff
    },
    "path/to/contractB.ext" : { ... },
    ...
  },
  "version" : "some-compiler-version-string"
}
```
It would be great if Vyper matched this output, at least as a default, so that it would enable easier integration with existing (and future) tools

## Motivation
My personal motivation is that I want to write a set of tools for Ethereum's Python ecosystem that make it easy to test, deploy, and interact with contracts. Some of the existing tools already leverage Solidity's output for this purpose (or otherwise hardcode themselves to solidity's style). Vyper, being the new kid on the block, should integrate with these tools, and I believe the best way of doing that is by using the same output (`--combined-json` in solidity) so that these tools can be useful.

My particular vision for a project written in this ecosystem would have the following folder structure:
```bash
project/
    contracts/
        contracts.json
        # arbitrarily nested contract files: .sol, .vy, etc.
    tests/  # pytest-ethereum: py.test plugin for contract testing
        config.yml  # References ../contracts.json, auto-generates pytest fixtures with given args
        # arbitrarily nested tests: test_*.py
    deployments/  # eth-deployer: enables easy deployment management
        config.yml  # References ../contracts.json, contains deployment args
        addresses.yml  # Deployment address tracking file
        # eth-deployer would have a console for interactions
        # Simulation and deployment files, etc.
    ...
    # Formal requirement verification, simulations, coverage testing, etc...
```
It doesn't have to look like this in the long run, my point is that it all builds off this common artifact output, and JSON is a good format to store these artifacts in. Later down the road, we can validate the compiler version present in the artifact output file and check some checksum hashes of the input files (stored in the compiler artifacts) and enable fast rebuilds when only certain files were changed in the `contracts/` directory (updating alongside similar output from solc). But I digress...

Adopting this as a standard (and this might require writing an explicit standard with the solidity team) would go to great lengths to have a compiler-neutral output format that any existing or new tools can be built around, where compiler artifacts (e.g. ABI, bytecode, etc.) can be inputs that are used easily and safely.

## Specification
If I had my way, I would change the syntax of the JSON file slightly, to something like this:
```python
{
  "contracts" : {
    "path/to/contractA.ext" : {
      "abi" : [ 
            ...  # No quotes necessary
       ],
      "bytecode" : "0x...",  # Changed name to 'bytecode'
      "runtime" : "0x...",  # Changed name to 'runtime', not sure how this gets used vs. bytecode
      ... # Additional stuff gets filtered out by tools
      # Suggestions:
      #      bytecode -> source code mapping (necessary for coverage frameworks)
      #      bytecode -> gas (may be necessary for gas profiling tools)
    },
    "path/to/contractB.ext" : { ... },
    ...
  },
  "version" : "some-compiler-version-string"
}
```

## Backwards Compatibility
I would say this is backward compatible, but would probably require a slight re-write of the docs:

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

**Comment by fubuloubu** [comment]
Implementation of this would be fairly straightforward, just a matter of creating a combined json output in `bin/viper`

**Comment by jacqueswww** [comment]
@fubuloubu How would we generate the contract name, if not from the filename? :stuck_out_tongue:

**Comment by fubuloubu** [comment]
Yeah, it would have to be the base of the filename. That's my wish list, I sort of wish it used the Contract Name directly instead of the filename:contract combo because it makes it easier to do the lookup. Solidity complains when there's overshadowing of the contract names anyways, so there's no reason to have the filenames be the unique key.

At least that would simplify the syntax of some of the things I'm working on: `contract_tester('path/to/File.v.py:File')` becomes `contract_tester('File')`

**Comment by yann300** [comment]
:+1:  for this, also the compiler input could be compliant with the standard used by solidity:
https://solidity.readthedocs.io/en/develop/using-the-compiler.html#input-description
And that would allow an easier integration to remix

**Comment by fubuloubu** [comment]
@yann300 Can you expand on the remix integration?

**Comment by yann300** [comment]
remix uses standard json i/o described in the link above, so that would not be so much of a problem to integrate a vyper compiler in it (well considering vyper compiler can be emscripted).

**Comment by fubuloubu** [comment]
Hmm. Very interesting. I found this: http://pypyjs.org/

Maybe there is a way to cross-compiler our language to JS
